src/style-spec/feature_filter/index.js

```diff
@@ -96,7 +94,19 @@ function compare(a, b) {
     return a < b ? -1 : a > b ? 1 : 0;
 }
 
+function geometryNeeded(filter) {
+    if (!Array.isArray(filter)) return false;
+    let needed = false;
+    for (let index = 0; index < filter.length; index++) {
+        const val = filter[index];
+        const op = Array.isArray(val) ? val[0] : val;
+        needed = needed || (op === 'within');
+    }
+    return needed;
```

Comment:
looks like this is a linear traversal of the first level of the filter, wouldn't that still break if the within expression is nested futher inside?

would DF/BF be better instead?

Reply:
Good point! Fixed, test is updated accordingly. Thank you!

Reply:
❤️ some clean recursion